### DISCUSSION TITLE
DateTime columns can not be used

### DISCUSSION THREAD

**Discussion by jiri-otoupal** [sqlalchemy/sqlalchemy#9990] [Usage Questions] [answered]
### Describe the bug

I have simple table with one column that has type DateTime, every time I want to query record that has date set it fails with exception

```
File "lib\sqlalchemy\cyextension\resultproxy.pyx", line 16, in sqlalchemy.cyextension.resultproxy.BaseRow.__init__
  File "lib\sqlalchemy\cyextension\resultproxy.pyx", line 73, in sqlalchemy.cyextension.resultproxy._apply_processors
  File "lib\sqlalchemy\cyextension\processors.pyx", line 34, in sqlalchemy.cyextension.processors.str_to_datetime
TypeError: fromisoformat: argument must be str
```

### SQLAlchemy Version in Use

2.0.2,2.0.16

### DBAPI (i.e. the database driver)

pysqlite

### Database Vendor and Major Version

SQLite

### Python Version

3.11

### Operating system

Windows

### To Reproduce

```python
Create table with datetime column, insert record with datetime, query with .all()
```

### Error

```
File "lib\sqlalchemy\cyextension\resultproxy.pyx", line 16, in sqlalchemy.cyextension.resultproxy.BaseRow.__init__
  File "lib\sqlalchemy\cyextension\resultproxy.pyx", line 73, in sqlalchemy.cyextension.resultproxy._apply_processors
  File "lib\sqlalchemy\cyextension\processors.pyx", line 34, in sqlalchemy.cyextension.processors.str_to_datetime
TypeError: fromisoformat: argument must be str
```

**Comment by CaselIT** [comment]
Hi,

Moving to discussion since there is no reproduction case.

**Reply by vehjt** [reply]
Im experiencing this exact error. 

Im using Langchain, which implements SQLAlchemy, to connect a Large Language Model to a already existing SQLite database. If i use a database that has a table with a column that has a TIMESTAMP type, i get these exact errors. If i use databases that use INT type as a timestamp, there are no issues.

**Reply by CaselIT** [reply]
I would report that to the library you are using if it's a library feature that's causing such issue.
Also this keeps not bing actionable, since there is no code that can be run. Can you create a reproducible example using only sqlalchemy?

**Reply by zzzeek** [reply]
SQLite does not really have "column types", it has "affinities" that are described on a per-row basis.  if your "existing database" has integers in columns, whether they are noted as the "TIMESTAMP" type or if they are noted as the "EPOCH_INT" type (which is what this sounds like), the values are not compatible with SQLAlchemy's SQLite TIMESTAMP type which is intended for ISO-formatted strings.    if you have epochs in those columns, see the `MyEpochType` example at https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator and use something like that.

**Reply by vehjt** [reply]
Im sorry for the half baked reply. Upon further expection of the mentioned database, it does not have anything to do with the TYPE of the column, but rather with the contents of the rows in that column. I have for example a database that stores this value in a TIMESTAMP column: 672457193.7343056 and this is what causes the error in SQLAlchemy

Hope this helps.

I do not exacly know how to use SQLAlchemy without Langchain, but i could figure it out and use it to collect the data from the databse with the mentioned TIMESTAMP column and expect to see the same results, due to the dot (.) in the timestamp itself.

**Reply by vehjt** [reply]
To recreate the error:
    
    
    from sqlalchemy import create_engine, Table, Column, TIMESTAMP, MetaData, select
    from sqlalchemy.orm import Session
    import sqlite3

    # Make new test database, create a timestamp table and column and insert value

    database_path = 'testdb.sqlite'
    timestamp_value = 672457193.7343056
    create_table_sql = """
    CREATE TABLE IF NOT EXISTS timestamps (
        timestamp TIMESTAMP
    );
    """

    insert_sql = """
    INSERT INTO timestamps (timestamp) VALUES (?);
    """

    conn = sqlite3.connect(database_path)
    cursor = conn.cursor()
    cursor.execute(create_table_sql)
    cursor.execute(insert_sql, (timestamp_value,))
    conn.commit()
    conn.close()

    print("The table has been created and the value has been inserted successfully.")



    #Use SQLAchemy to fetch all results from the timestamp table and recreate the error.

    database_path = 'testdb.sqlite'
    table_name = 'timestamps'  # Replace 'your_table_name' with the name of your table
    engine = create_engine(f'sqlite:///{database_path}')
    metadata = MetaData()
    metadata.reflect(bind=engine)
    reflected_table = metadata.tables[table_name]

    with Session(engine) as session:
        result = session.execute(select(reflected_table)).all()

        # Print the results
        for row in result:
            print(row)

**Reply by CaselIT** [reply]
As mentioned sqlalchemy doesn't support timestamp columns in sqlite that contain numbers. Use another data type for it instead

**Comment by CaselIT** [comment]
Also I cannot reproduce
```py
import sqlalchemy as sa
import datetime as dt

t = sa.Table("t", sa.MetaData(), sa.Column("c", sa.DateTime()))
e = sa.create_engine("sqlite://")
t.metadata.create_all(e)
with e.connect() as c:
    c.execute(t.insert(), [{"c": dt.datetime.now()}, {"c": None}])
    print(c.execute(t.select()).all())
```

That's why the template asks for a reproduction code

**Comment by zzzeek** [comment, marked as answer]
Here is the answer to this question:


SQLite does not have real column types.  it instead has a notion of [type affinities](https://www.sqlite.org/datatype3.html) , a loose association of naming conventions to a **suggestion** for what kind of data might be in a column.  However, **any type of data can be in any column in any row in SQLite, and declared "type" has nothing to do with this whatsoever** (keep this in mind next time you read one of these "SQLite is so amazingly well architected and designed!" threads :roll_eyes: ).

SQLAlchemy's date/time/datetime types for SQLite represent date/time/datetime data in the database using [only strings of a specific, pre-determined format](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME).  This format defaults to the following [ISO strftime formats](https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior):

DATETIME: `"%(year)04d-%(month)02d-%(day)02d %(hour)02d:%(minute)02d:%(second)02d.%(microsecond)06d"`
DATE: `"%(year)04d-%(month)02d-%(day)02d"`
TIME: `"%(hour)02d:%(minute)02d:%(second)02d.%(microsecond)06d"`

if your date/time/timestamp are in a **different** string format, you [must customize both the string format and the regular expression used to parse them](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME).

if your date/time/timestamp are **not strings, but actually integers**, you cannot use these datatypes; use a type like the [epoch datatype](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator):

```py

class MyEpochType(types.TypeDecorator):
    impl = types.Integer
    cache_ok = True

    epoch = datetime.date(1970, 1, 1)

    def process_bind_param(self, value, dialect):
        return (value - self.epoch).days

    def process_result_value(self, value, dialect):
        return self.epoch + timedelta(days=value)
```

So to sum up:

1. SQLite doesnt have real "types", it has only loosely defined "type affinities"
2. SQLite makes no requirement on what kind of data should be present with a certain "type"
3. SQLAlchemy's SQLite date/time/datetime datatypes use strings in a specific ISO format
4. the format can be customized but has to be a string
5. integers are often used for "TIMESTAMP" types of data, use an "epoch" style type if this is what your database has

**Reply by CaselIT** [reply]
recentish version of sqlite have a strict mode that do actually enforce typing (I think)
support for that is tracked at https://github.com/sqlalchemy/sqlalchemy/issues/7398

**Reply by vehjt** [reply]
Thank you for the reply. In my context im not actualy sure what to do next, since SQLAlchemy is implemented by Langchain's SQL-agent. I think ill have to go to them and point this out somehow.
